geoengine/raster.py

```diff
@@ -212,3 +212,93 @@ def from_ge_record_batch(record_batch: pa.RecordBatch) -> RasterTile2D:
             time,
             band,
         )
+
+
+class RasterTileStack2D:
+    '''A stack of all the bands of a raster tile as produced by the Geo Engine'''
+    size_y: int
+    size_x: int
+    geo_transform: gety.GeoTransform
+    crs: str
+    time: gety.TimeInterval
+    data: List[pa.Array]
+    bands: List[int]
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+            self,
+            tile_shape: Tuple[int, int],
+            data: List[pa.Array],
+            geo_transform: gety.GeoTransform,
+            crs: str,
+            time: gety.TimeInterval,
+            bands: List[int],
+    ):
+        '''Create a RasterTileStack2D object'''
+        (self.size_y, self.size_x) = tile_shape
+        self.data = data
+        self.geo_transform = geo_transform
+        self.crs = crs
+        self.time = time
+        self.bands = bands
+
+    def single_band(self, index: int) -> RasterTile2D:
+        '''Return a single band from the stack'''
+        return RasterTile2D(
+            (self.size_y, self.size_x),
+            self.data[index],
+            self.geo_transform,
+            self.crs,
+            self.time,
+            self.bands[index],
+        )
+
+    def to_numpy_masked_array_stack(self) -> np.ma.MaskedArray:
+        '''Return the raster stack as a 3D masked numpy array'''
+        arrays = [self.single_band(i).to_numpy_masked_array() for i in range(0, len(self.data))]
+        stack = np.stack(arrays, axis=0)
+        return stack
+
+    def to_xarray(self, clip_with_bounds: Optional[gety.SpatialBounds] = None) -> xr.DataArray:
+        '''Return the raster stack as an xarray.DataArray'''
+        arrays = [self.single_band(i).to_xarray(clip_with_bounds) for i in range(0, len(self.data))]
+        stack = xr.concat(arrays, dim='band')
+        return stack
+
+
+async def tile_stream_to_np_stack(raster_stream: AsyncIterator[RasterTile2D]) -> AsyncIterator[RasterTileStack2D]:
+
+    ''' Convert a stream of raster tiles to stream of stacked tiles '''
+    store: List[RasterTile2D] = []
+    first_band: int = -1
+
+    async for tile in raster_stream:
```

Comment:
would it make sense to do some sanity checks here that the correct tiles arrived? Or should be trust the backend?

Reply:
[a7db320](https://github.com/geo-engine/geoengine-python/pull/172/commits/a7db3207e3997c866e28dfede7b439fe7b7ccb48)

then had to change more things:

[ad0bca8](https://github.com/geo-engine/geoengine-python/pull/172/commits/ad0bca8e154a75444fc57695e503b6f57b192120)